src/game/common/statemachine.cpp

```diff
@@ -0,0 +1,716 @@
+{
+    if (!m_locked) {
+        return Internal_Set_State(new_state_id);
+    } else {
+#ifdef GAME_DEBUG_STRUCTS
+        if (m_currentState != nullptr) {
+            captainslog_debug(" cur state '%s'", m_currentState->Get_Name().Str());
+        }
+
+        captainslog_debug("machine is locked (by %s), cannot be cleared (Please don't ignore; this generally indicates a "
+                          "potential logic flaw)",
+            m_lockedString);
+#endif
+        return STATE_CONTINUE;
+    }
+}
+
+StateReturnType StateMachine::Internal_Set_State(unsigned int new_state_id)
+{
+    State *new_state = nullptr;
+    m_updateFrame = 0;
+
+    if (new_state_id != MACHINE_DONE_STATE_ID) {
+        if (new_state_id == INVALID_STATE_ID) {
+            new_state_id = m_defaultStateID;
+
+            if (new_state_id == INVALID_STATE_ID) {
+                captainslog_debug("you may NEVER set the current state to an invalid state id.");
+                return STATE_FAILURE;
+            }
+        }
+
+        new_state = Internal_Get_State(new_state_id);
+#ifdef GAME_DEBUG_STRUCTS
+        if (Get_Wants_Debug_Output()) {
+            unsigned int id = INVALID_STATE_ID;
+
+            if (m_currentState != nullptr) {
+                id = m_currentState->Get_ID();
+            }
+
+            captainslog_debug(
+                "%d '%s'%x -- '%s' %x exit ", g_theGameLogic->Get_Frame(), m_name.Str(), m_owner, m_name.Str(), this);
+
+            if (m_currentState != nullptr) {
+                captainslog_debug(" '%s' ", m_currentState->Get_Name().Str());
+            } else {
+                captainslog_debug(" INVALID_STATE_ID ");
+            }
+
+            if (new_state != nullptr) {
+                captainslog_debug("enter '%s' ", new_state->Get_Name().Str());
+            } else {
+                captainslog_debug("to INVALID_STATE");
+            }
+        }
+#endif
+    }
+
+    if (m_currentState != nullptr) {
+        m_currentState->On_Exit(EXIT_NORMAL);
+    }
+
+    m_currentState = new_state;
+
+    if (m_currentState == nullptr) {
+        return STATE_CONTINUE;
+    }
+
+    State *old_state = m_currentState;
+    StateReturnType state = m_currentState->On_Enter();
+
+    if (!m_currentState) {
+        return STATE_FAILURE;
+    }
+
+    if (old_state != m_currentState) {
+        state = STATE_CONTINUE;
+    }
+
+    if (state <= STATE_CONTINUE) {
+        return m_currentState->Friend_Check_For_Transitions(state);
+    }
+
+    unsigned int frame = g_theGameLogic->Get_Frame();
+    m_updateFrame = state + frame;
+    return m_currentState->Friend_Check_For_Sleep_Transitions((StateReturnType)(m_updateFrame - frame));
+}
+
+StateReturnType StateMachine::Init_Default_State()
+{
+#ifdef GAME_DEBUG_STRUCTS
+    for (auto &iter : m_stateMap) {
+        State *state = iter.second;
+        unsigned int id = state->Get_ID();
+        std::vector<unsigned int> *vector = state->Get_ID_Vector();
+
+        if (!vector->empty()) {

```

Comment:
This condition is obsolete and can be removed.

Reply:
fixed.